tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,747 @@
+                EntryState::Cancelled => panic!("Trying to register a cancelled timer"),
+                EntryState::FiringInProgress => panic!("Unexpected FiringInProgress state"),
+                EntryState::Fired | EntryState::Registered => (), // ok - reregistration
+                EntryState::NotRegistered => (),
+            };
+
+            self.inner
+                .as_ref()
+                .state
+                .set(EntryState::Registered, Ordering::Relaxed);
+
+            true
         }
     }
 
-    fn new2(deadline: Instant, duration: Duration, inner: Weak<Inner>, state: u64) -> Self {
-        Self {
-            time: CachePadded(UnsafeCell::new(Time { deadline, duration })),
-            inner,
-            waker: AtomicWaker::new(),
-            state: AtomicU64::new(state),
-            queued: AtomicBool::new(false),
-            error: AtomicU8::new(0),
-            next_atomic: UnsafeCell::new(ptr::null_mut()),
-            when: UnsafeCell::new(None),
-            next_stack: UnsafeCell::new(None),
-            prev_stack: UnsafeCell::new(ptr::null_mut()),
+    /// Transitions to the FiringInProgress state
+    pub(super) unsafe fn set_pending(&self) {
+        unsafe {
+            assert_eq!(
+                self.inner.as_ref().state.get(Ordering::Relaxed),
+                EntryState::Registered
+            );
+            self.inner
+                .as_ref()
+                .state
+                .set(EntryState::FiringInProgress, Ordering::Relaxed);
         }
     }
 
-    fn upgrade_inner(&self) -> Option<Arc<Inner>> {
-        self.inner.upgrade()
-    }
-}
+    /// Attempts to transition to a terminal state. If the state is already a
+    /// terminal state, does nothing.
+    ///
+    /// Because the entry might be dropped after the state is moved to a
+    /// terminal state, this function consumes the handle to ensure we don't
+    /// access the entry afterwards.
+    ///
+    /// Returns the last-registered waker, if any.
+    ///
+    /// SAFETY: The driver lock must be held while invoking this function.
+    pub(super) unsafe fn fire(self, completed_state: EntryState) -> Option<Waker> {
+        debug_assert!(completed_state.is_elapsed());
+
+        // Get a lifetime-erased pointer to the state. We'll use this to set the
+        // elapsed state below.
+        let pstate = unsafe { self.inner.as_ref() }.state.erase();
+
+        // If we're already in a terminal state, do nothing. This typically
+        // happens when trying to cancel a timer that's already fired.
+        //
+        // Ordering: State is changed only under the driver lock.
+        match pstate.get(Ordering::Relaxed) {
+            EntryState::FiringInProgress | EntryState::NotRegistered | EntryState::Registered => { /* continue */
+            }
+            EntryState::Fired | EntryState::Cancelled | EntryState::Error(_) => {
+                return None;
+            }
+        };
```

Comment:
nit/take it or leave it: it might be a little more concise to just write
```suggestion
        if let EntryState::Fired | EntryState::Cancelled | EntryState::Error(_) = pstate.get(Ordering::Relaxed) {
            return None;
        }
```